source/postprocess/visualization/density_anomaly.cc

```diff
@@ -0,0 +1,246 @@
+/*
+  Copyright (C) 2011 - 2024 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/postprocess/visualization/density_anomaly.h>
+#include <aspect/adiabatic_conditions/interface.h>
+#include <aspect/lateral_averaging.h>
+#include <aspect/utilities.h>
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+
+
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+      template <int dim>
+      DensityAnomaly<dim>::
+      DensityAnomaly ()
+        :
+        CellDataVectorCreator<dim>("kg/m^3")
+      {}
+
+
+
+      template <int dim>
+      std::pair<std::string, std::unique_ptr<Vector<float>>>
+      DensityAnomaly<dim>::execute() const
+      {
+        std::pair<std::string, std::unique_ptr<Vector<float>>>
+        return_value ("density_anomaly",
+                      std::make_unique<Vector<float>>(this->get_triangulation().n_active_cells()));
+
+        // The following lines evaluate the density at a single point per cell
+        // and then compute the difference from the average in
+        // the slice within which the point lies
+        const QMidpoint<dim> quadrature_formula;
+        const unsigned int n_q_points = quadrature_formula.size();
+
+
+        FEValues<dim> fe_values (this->get_mapping(),
+                                 this->get_fe(),
+                                 quadrature_formula,
+                                 update_values   |
+                                 update_gradients |
+                                 update_quadrature_points );
+
+        MaterialModel::MaterialModelInputs<dim> in(n_q_points, this->n_compositional_fields());
+        MaterialModel::MaterialModelOutputs<dim> out(n_q_points, this->n_compositional_fields());
+
+        switch (average_density_scheme)
+          {
+            case reference_profile:
+            {
+              // Loop over the cells
+              for (const auto &cell : this->get_dof_handler().active_cell_iterators())
+                if (cell->is_locally_owned())
```

Comment:
Did I not realize this before, or did you switch the order of this logic? Either way, could you please reorder the code in this plugin in the following structure:
```
if (average_density_scheme == lateral_average)
{
   ... do all the preparation for lateral averages
}

for (const auto &cell : this->get_dof_handler().active_cell_iterators())
{
  do all the common things, like fe_values.reinit and material model evaluate

  switch (average_density_scheme)
  ... here put all the things that are different between reference_profile and lateral_average
}
```

Currently you duplicate a lot of code between the two options, but with my structure you will need less code. Sorry for the work, but less code duplication means less work and less bugs in the future.